In [78]:
import pandas as pd


pd.set_option('display.max_rows', None)

### Carregar arquivos no Pandas

In [79]:
cartoes = pd.read_csv("campeonato-brasileiro-cartoes.csv")
estatisticas_full = pd.read_csv("campeonato-brasileiro-estatisticas-full.csv")
full = pd.read_csv("campeonato-brasileiro-full.csv")
gols = pd.read_csv("campeonato-brasileiro-gols.csv")

Vamos renomar as colunas

In [80]:
cartoes = cartoes.rename(columns={
    'rodata': 'rodada'
})
estatisticas_full = estatisticas_full.rename(columns={
    'rodata': 'rodada'
})
full = full.rename(columns={
    'rodata': 'rodada'
})
gols = gols.rename(columns={
    'rodata': 'rodada'
})

## Limpeza do dataframe cartoes

#### Verificar inconsistencias nas colunas

In [81]:
print(cartoes.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 18857 entries, 0 to 18856
Data columns (total 8 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   partida_id  18857 non-null  int64  
 1   rodada      18857 non-null  int64  
 2   clube       18857 non-null  object 
 3   cartao      18857 non-null  object 
 4   atleta      18851 non-null  object 
 5   num_camisa  18471 non-null  float64
 6   posicao     17659 non-null  object 
 7   minuto      18857 non-null  object 
dtypes: float64(1), int64(2), object(5)
memory usage: 1.2+ MB
None


Verificar as colunas que tem valor nulo

In [82]:
print(cartoes.isna().sum())

partida_id       0
rodada           0
clube            0
cartao           0
atleta           6
num_camisa     386
posicao       1198
minuto           0
dtype: int64


Verificando quais sao os atletas com valor nulo:

In [83]:

indices_nulos = cartoes[cartoes['atleta'].isna()].index.tolist()
# Exibir as linhas correspondentes
print(cartoes.loc[indices_nulos])

       partida_id  rodada          clube    cartao atleta  num_camisa  \
8471         6303      20        Vitoria   Amarelo    NaN        14.0   
8966         6420      30        Vitoria   Amarelo    NaN        14.0   
12359        7124      24  Internacional   Amarelo    NaN        14.0   
13106        7291       3  Internacional  Vermelho    NaN        14.0   
13174        7309       5  Internacional   Amarelo    NaN        14.0   
13281        7332       7  Internacional   Amarelo    NaN        14.0   

        posicao minuto  
8471   Zagueiro     63  
8966   Zagueiro     34  
12359  Zagueiro     58  
13106  Zagueiro     52  
13174  Zagueiro     65  
13281  Zagueiro     37  


Tentando achar os valores nulos em outras partidas.

Vou verificar por meio do time + numero da camisa para encontrar a posicao que falta

In [84]:
resultado = cartoes[(cartoes['clube'].fillna('') == 'Vitoria') & (cartoes['num_camisa'].fillna(0) == 14)]
print(resultado)

      partida_id  rodada    clube    cartao         atleta  num_camisa  \
189         4646       4  Vitoria   Amarelo   José Welison        14.0   
479         4715      10  Vitoria   Amarelo   José Welison        14.0   
515         4723      11  Vitoria   Amarelo   José Welison        14.0   
652         4756      14  Vitoria   Amarelo   José Welison        14.0   
986         4824      21  Vitoria   Amarelo   José Welison        14.0   
1603        4941      35  Vitoria   Amarelo   José Welison        14.0   
3853        5382       2  Vitoria   Amarelo   José Welison        14.0   
4549        5525      16  Vitoria   Amarelo   José Welison        14.0   
5288        5680      32  Vitoria   Amarelo   José Welison        14.0   
7169        6072      33  Vitoria   Amarelo   José Welison        14.0   
7275        6092      35  Vitoria   Amarelo   José Welison        14.0   
7820        6197       8  Vitoria   Amarelo   José Welison        14.0   
8471        6303      20  Vitoria   Am

Verificamos que o atleta do vitoria com a camisa 14 e Zagueiro eh o Lucas Ribeiro.


In [85]:
cartoes.loc[[8966], 'atleta'] = 'Lucas Ribeiro'
cartoes.loc[[8471], 'atleta'] = 'Lucas Ribeiro'

Agora vamos verificar quem eh o Zagueiro do Inter camisa 14

In [86]:
resultado = cartoes[(cartoes['clube'].fillna('') == 'Internacional') & (cartoes['num_camisa'].fillna(0) == 14)]

[Partida para verificar quem tomou o cartao](https://www.google.com/search?q=06%2F12%2F2020+Atletico-MG%22%2C%22Internacional&oq=06%2F12%2F2020+Atletico-MG%22%2C%22Internacional&gs_lcrp=EgZjaHJvbWUyBggAEEUYOTIGCAEQLhhA0gEINTQ1NmowajGoAgCwAgA&sourceid=chrome&ie=UTF-8#sie=m;/g/11hf7l54j1;2;/m/0fnk7q;tl;fp;1;;;)

Logo, os valores nulos dos cartoes sao do Lucas Ribeiro

In [87]:
cartoes.loc[[12359], 'atleta'] = 'Lucas Ribeiro'
cartoes.loc[[13106], 'atleta'] = 'Lucas Ribeiro'
cartoes.loc[[13174], 'atleta'] = 'Lucas Ribeiro'
cartoes.loc[[13281], 'atleta'] = 'Lucas Ribeiro'

In [88]:
indices_nulos = cartoes[cartoes['atleta'].isna()].index.tolist()
# Exibir as linhas correspondentes
print(cartoes.loc[indices_nulos])

Empty DataFrame
Columns: [partida_id, rodada, clube, cartao, atleta, num_camisa, posicao, minuto]
Index: []


## Agora vamos verificar quais posições dos atletas estão vazio

In [89]:
pos_vazio = cartoes[cartoes["posicao"].isna()]
print(pos_vazio.info())

<class 'pandas.core.frame.DataFrame'>
Index: 1198 entries, 0 to 9431
Data columns (total 8 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   partida_id  1198 non-null   int64  
 1   rodada      1198 non-null   int64  
 2   clube       1198 non-null   object 
 3   cartao      1198 non-null   object 
 4   atleta      1198 non-null   object 
 5   num_camisa  1140 non-null   float64
 6   posicao     0 non-null      object 
 7   minuto      1198 non-null   object 
dtypes: float64(1), int64(2), object(5)
memory usage: 84.2+ KB
None


Nesse caso vou fazer uma função para verificar quais atletas já apareceram no banco de dados e se essas outras linhas tem o valor posicao para poder preencher as que não tem.

In [90]:
def verificar_atleta_duplicado(cartoes, linhas_com_nan) -> pd.DataFrame:
    """
    Para cada linha de linhas_com_nan, verifica se o valor da coluna 'atleta' aparece mais de uma vez no DataFrame df.
    
    Retorna um novo DataFrame (cópia de linhas_com_nan) com uma coluna extra 'atleta_duplicado'
    que indica com True se o atleta aparece em outra linha e False caso contrário.
    """
    # Cria uma cópia para não alterar o DataFrame original
    resultado = linhas_com_nan.copy()
    
    # Para cada atleta, conta quantas vezes ele aparece no df
    # e verifica se o número de ocorrências é maior que 1.
    resultado['atleta_duplicado'] = resultado['atleta'].apply(
        lambda atleta: (cartoes['atleta'] == atleta).sum() > 1
    )
    
    return resultado

resultado = verificar_atleta_duplicado(cartoes, pos_vazio)

Agora vamos de fato trocar os valores NaN para as outras posições correspondentes 

In [91]:
def preencher_posicao_na_duplicatas(df):
    """
    Para cada atleta que aparece mais de uma vez no DataFrame,
    verifica se existe uma linha com um valor válido (não-NaN) na coluna 'posicao'.
    Se existir, preenche os valores NaN de 'posicao' para esse atleta com o valor encontrado.
    
    Retorna um novo DataFrame com as atualizações.
    """
    df_atualizado = df.copy()
    
    # Agrupar por atleta
    for atleta, group in df_atualizado.groupby('atleta'):
        # Considera somente atletas duplicados
        if group.shape[0] > 1:
            # Procura posições não-NaN no grupo
            posicoes_validas = group['posicao'].dropna().unique()
            if len(posicoes_validas) > 0:
                # Atualiza todas as linhas desse atleta onde 'posicao' é NaN
                df_atualizado.loc[
                    (df_atualizado['atleta'] == atleta) & (df_atualizado['posicao'].isna()),
                    'posicao'
                ] = posicoes_validas[0]
    return df_atualizado


df_atualizado = preencher_posicao_na_duplicatas(cartoes)


In [92]:
pos_vazio = df_atualizado[df_atualizado["posicao"].isna()]
print(pos_vazio.info())

<class 'pandas.core.frame.DataFrame'>
Index: 144 entries, 45 to 9276
Data columns (total 8 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   partida_id  144 non-null    int64  
 1   rodada      144 non-null    int64  
 2   clube       144 non-null    object 
 3   cartao      144 non-null    object 
 4   atleta      144 non-null    object 
 5   num_camisa  130 non-null    float64
 6   posicao     0 non-null      object 
 7   minuto      144 non-null    object 
dtypes: float64(1), int64(2), object(5)
memory usage: 10.1+ KB
None


### Importante verificar que saimos de 1198 valores com posição nula para 144 apenas!

### Agora vamos normalizar os valores de posicao

In [93]:
cartoes['num_camisa'] = cartoes['num_camisa'].fillna(0).astype(int)
print(cartoes['num_camisa'].info())

<class 'pandas.core.series.Series'>
RangeIndex: 18857 entries, 0 to 18856
Series name: num_camisa
Non-Null Count  Dtype
--------------  -----
18857 non-null  int64
dtypes: int64(1)
memory usage: 147.4 KB
None


### Cartoes ja estao normalizados, vamos agora para estatisticas full

In [94]:
print(estatisticas_full['partida_id'].isna().sum()) # Sem partida com valor nulo


0


Precisao de passes e posse de bola estao em porcentagem

Vou transformar para int e transformar valores NaN em -1


In [95]:
estatisticas_full['posse_de_bola'] = estatisticas_full['posse_de_bola'].str.replace('%','').fillna(-1).astype(int)
estatisticas_full['precisao_passes'] = estatisticas_full['precisao_passes'].str.replace('%','').fillna(-1).astype(int)
estatisticas_full.tail(3)

partida_id  rodada       clube  chutes  chutes_no_alvo  posse_de_bola  \
16807        8404      38    Flamengo      10               2             58   
16808        8405      38       Vasco      13               6             45   
16809        8405      38  Bragantino       9               2             55   

       passes  precisao_passes  faltas  cartao_amarelo  cartao_vermelho  \
16807     535               89      16               3                0   
16808     321               78       8               3                0   
16809     398               80      14               1                1   

       impedimentos  escanteios  
16807             0           9  
16808             0           2  
16809             1           3

# Agora vamos normalizar Brasileiro Full

In [ ]:
full.fillna({
    'formacao_mandante': 'Desconhecida',
    'formacao_visitante': 'Desconhecida',
    'tecnico_mandante': 'Não informado',
    'tecnico_visitante': 'Não informado'
}, inplace=True)
full


# Exportar para SQL

In [ ]:
from sqlalchemy import create_engine


engine = create_engine('postgresql://usuario:senha@localhost:5432/meu_banco')


cartoes.to_sql('cartoes', engine, if_exists='replace', index=False)
estatisticas_full.to_sql('estatisticas_full', engine, if_exists='replace', index=False)
full.to_sql('full', engine, if_exists='replace', index=False)
gols.to_sql('gols', engine, if_exists='replace', index=False)